crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,248 @@
+        return Err(result);
+    }
+    Ok(())
+}
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
+    let c_mode_string = CString::new(so_name).expect("modeString CString::new failed");
+    let cptr_mode_string = c_mode_string.as_ptr();
+
+    let mut get_interface_func: nvtxw_bindings::nvtxwGetInterface_t = Default::default();
+    check(unsafe {
+        nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            cptr_mode_string,
+            &mut get_interface_func,
+            null_mut(),
+        )
+    })?;
+
+    let get_interface = get_interface_func.ok_or(NVTXW3_RESULT_FAILED)?;
+
+    let mut interface_void: *const c_void = null();
+    check(unsafe { get_interface(NVTXW3_INTERFACE_ID_CORE_V1, &mut interface_void) })?;
+
+    if interface_void.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    let ptr_interface = interface_void as *const nvtxw_bindings::nvtxwInterfaceCore_t;
+    let interface;
+    unsafe {
+        interface = *ptr_interface;
+    }
+
+    Ok(interface)
+}
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    o_output: Option<PathBuf>,
+    o_merge: Option<PathBuf>,
+) -> Result<SessionHandle, ResultCode> {
+    let func = interface
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let mut c_output: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_output) = o_output {
+        let s_output = os_output.to_string_lossy();
+        let session_name = match s_output.strip_suffix(".nsys-rep") {
+            Some(s) => s,
+            None => &s_output,
+        };
+        c_output = CString::new(session_name).expect("CString::new failed");
+    }
+
+    let mut c_config: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_merge) = o_merge {
+        let s_merge = os_merge.to_string_lossy();
+        let s_config = format!("ReportMerge={}", s_merge);
+
+        c_config = CString::new(s_config).expect("CString::new failed");
+    }
+
+    let session_attr = SessionAttributes {
+        struct_size: size_of::<SessionAttributes>(),
+        name: if c_output.is_empty() {
+            null_mut()
+        } else {
+            c_output.as_ptr()
+        },
+        configString: if c_config.is_empty() {
+            null_mut()
+        } else {
+            c_config.as_ptr()
+        },
+    };
+
+    let mut session: SessionHandle = Default::default();
+    check(unsafe { func(&mut session, &session_attr) })?;
+
+    if session.opaque.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    Ok(session)
+}
+
+pub fn stream_open_simple(
+    interface: InterfaceHandle,
+    session: SessionHandle,
+    stream_name: String,
+    domain_name: String,
+) -> Result<StreamHandle, ResultCode> {
+    let func = interface.StreamOpen.ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
```

Comment:
If it were me I'd name this something more descriptive like `stream_open`, so that when you go call it down below it's obvious what you're calling. I get that this follows a pattern and some of the later API calls are shorter (thus making it more obvious), but in these longer API calls, you have to read back a bit to figure out what is being called.

Reply:
Done